pkg/controller/metrics.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/exposure-notifications-verification-server/pkg/observability"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+const contextKeyMetrics = contextKey("controllerMetricsKey")
+
+var MetricPrefix = observability.MetricRoot + "/server/shared"
+
+type Metrics struct {
+	FirebaseRecreates *stats.Int64Measure
+}
+
+func MetricsFromContext(ctx context.Context) (context.Context, *Metrics, error) {
```

Comment:
i don't love the dependency through context - it now implies an ordering between controller creation.

I still think there are other solutions that don't require that ordering / dependency

Reply:
What do you mean by ordering? This is lazily initialized as a singleton on context, if the controller can't find it on context, it gets created. Controllers can be organized in any order (the first one will create)

Reply:
it requires that everyone who calls it shares a context though, otherwise it will be created twice.

that may be OK in opencensus - I'm not positive

Reply:
I think OC actually requires you to use the ctx - all the data is stored on the ctx